File: back-end/src/main/java/com/huyphan/dtos/SearchCriteriaDto.java
package com.huyphan.dtos;

import com.huyphan.models.enums.ConstructorName;
import com.huyphan.models.enums.SearchCriteriaName;
import lombok.Getter;
import lombok.Setter;

/** Search criteria DTO. */
@Getter
@Setter
public class SearchCriteriaDto {

	/** Criteria name. */
	private SearchCriteriaName name;

	/** Construct which will use this search criteria to constructs Predicate. */
	private ConstructorName constructorName;

	/** Criteria value. */
	private Object value;
}

File: front-end/src/app/core/services/dtos/search-criteria-dto.ts
import { SearchCriteriaName } from '../../enums/search-criteria-name';
import { SearchCriteriaUsageDescription } from '../../enums/search-criteria-usage-description';

type SearchCriteriaDtoConstructorData<T> = Omit<SearchCriteriaDto<T>, 'shouldInclude'>;

/** Search criteria DTO. */
export abstract class SearchCriteriaDto<T> {
  /** Search criteria name. */
  public readonly name: SearchCriteriaName;

  /** Search criteria value. */
  public readonly value: T;

  /** Property required by the back-end.
   * A description of what this search criteria used for. This property will be used by the back-end to create search query dynamically. */
  public readonly constructorName: SearchCriteriaUsageDescription;

  public constructor(data: SearchCriteriaDtoConstructorData<T>) {
    this.name = data.name;
    this.value = data.value;
    this.constructorName = data.constructorName;
  }

  /** Determine whether to include this criteria in the criteria array or not. */
  public abstract shouldInclude(): boolean;
}
